### PR TITLE
warn if asyncio test requests async pytest fixture in strict mode

### PR DESCRIPTION
**Pull request by jakkdl**
partially fixes https://github.com/pytest-dev/pytest/issues/10839, see https://github.com/pytest-dev/pytest/issues/10839#issuecomment-2355427134

It got a bit messy to implement, and relies on non-public API in `_fixtureinfo`, so I'm not 100% that this is the best way to approach it. Another approach would be to try and catch this on collection, but then you end up having to handle overriding fixtures and the like.

It's also *possible* that this breaks tests for end users, but if they really do want an unawaited object from a fixture they can resolve it in various ways by returning an `Awaitable` from their fixture.

### CODE REVIEW

**Comment by jakkdl** [comment]
actually, it's not that unlikely that this will break tests for some users if they're using `@pytest.fixture(autouse=True)` and mixing asyncio and async-but-not-asyncio tests. But... that's probably not a great thing to do.

File: pytest_asyncio/plugin.py
Comment (jakkdl) [reply pytest_asyncio/plugin.py:893]: yeah that's probably good, I'll change it :)

**Comment by seifertm** [comment]
This looks great, thank you!

You mentioned in https://github.com/pytest-dev/pytest/pull/12930#issuecomment-2468341786 that this change may become redundant once the same check is in place in pytest. I'm generally okay with having the check in pytest-asyncio, too, at least temporarily. At some point, pytest-asyncio will bump the minimum required pytest version to a version where this check is included in pytest and we can remove it here.

I'd like to wait for a decision on your pytest PR before merging (whether you go with a deprecation or plain error), but other than that it looks good to me. 

Thanks for the effort!

**Comment by jakkdl** [comment]
Looks like we're going with a warning in the upstream PR.
It's probably good to have the check in pytest-asyncio, even if somewhat redundant, since for this case we can give a more specific & actionable error message for the user.

**Comment by seifertm** [comment]
I adjusted the release date of v0.25.0 to UNRELEASED and squashed the commits.
Thanks again @jakkdl !

**Review by seifertm — APPROVED** [review]

**Comment by endrift** [comment]
Soooooo...how am I supposed to get the semantics that this now-deprecated approach provided on modern pytest-asyncio? Previously, when in strict mode, I could mark it as `@pytest.fixture` and I'd get an Awaitable in the test. If the fixture were anything other than function scope this would probably be bad, but these fixtures are designed to be function scope. If I switch to `@pytest_asyncio.fixture`, as recommended, the semantics are different: the fixture has been awaited before the test starts. This is the same behavior in auto mode, but I wasn't using auto mode. I have yet to figure out a way to get those same semantics in a way that doesn't give a warning except for wrapping the bodies of all of my fixtures with an async def.

**Comment by jakkdl** [comment]
> Soooooo...how am I supposed to get the semantics that this now-deprecated approach provided on modern pytest-asyncio? Previously, when in strict mode, I could mark it as `@pytest.fixture` and I'd get an Awaitable in the test. If the fixture were anything other than function scope this would probably be bad, but these fixtures are designed to be function scope. If I switch to `@pytest_asyncio.fixture`, as recommended, the semantics are different: the fixture has been awaited before the test starts. This is the same behavior in auto mode, but I wasn't using auto mode. I have yet to figure out a way to get those same semantics in a way that doesn't give a warning except for wrapping the bodies of all of my fixtures with an async def.

fixtures are intended to be used for setup/teardown of tests, so if you're using it to provide awaitables to your tests my first instinct is that you should perhaps rethink the structure of your tests - make the fixtures standalone helper functions, or if the fixtures require additional setup before they can be run then make them depend on some other fixtures.

But if you simply do want to suppress it you indeed do need to wrap it (though you can wrap in a sync func), as documented in the yet-unreleased pytest 8.4 docs https://docs.pytest.org/en/latest/deprecations.html#sync-test-depending-on-async-fixture
If you're doing it a lot you can write a decorator as well
```python
def raw_awaitable(async_fixture):
    def inner():
        return async_fixture()
    return inner

@pytest.fixture
@raw_awaitable
async def my_fixture():
    return 5
```

**Comment by endrift** [comment]
The problem in my case is that my weird fixtures pull in other fixtures--mostly monkeypatch. I think I can restructure it so I just pass that as an argument to these instead of being fixtures themselves, but it was a lot of warnings so I haven't had the chance to go over all of it yet. I think at least one of them does teardown at the end though. I have to keep going through the list

**Comment by endrift** [comment]
Looks like none of them did teardown, so just replacing the fixture usages with calls worked.

**Pull request merged**
